### PR TITLE
Warn when cookie domain is set to an IP

### PR DESCRIPTION
**Pull request by JJPO96**
Possible solution for issue #2007, it gives a warning message when an IP is set as SESSION_COOKIE_DOMAIN as user untitaker asked us to do, SERVER_DOMAIN isn't checked because he didn't remembered why it was needed, but the function is_IP in /flask/helpers is prepared to it. If you have any doubt feel free to ask.

### CODE REVIEW

File: flask/helpers.py [hunk 1]
```diff
@@ -958,3 +958,33 @@ def total_seconds(td):
     :rtype: int
     """
     return td.days * 60 * 60 * 24 + td.seconds
+
+def is_IP(string, var_name):
```
Comment (ThiefMaster) [inline flask/helpers.py]: PEP8: should be `is_ip`

Also, the naming is very different from what the function does.

File: flask/sessions.py [hunk 2]
```diff
@@ -332,6 +332,7 @@ def open_session(self, app, request):
 
     def save_session(self, app, session, response):
         domain = self.get_cookie_domain(app)
+        is_IP(domain, "SESSION_COOKIE_DOMAIN")
```
Comment (ThiefMaster) [inline flask/sessions.py]: ideally you would use `if is_ip(domain): warnings.warn(...)` instead of having the function print something

**Comment by JJPO96** [comment]
What name would you suggest? Regards

**Comment by ThiefMaster** [comment]
I think instead of string operations the proper way to check whether it's an IP is using [`socket.inet_pton()`](https://docs.python.org/2/library/socket.html#socket.inet_pton) and returning `True` if it doesn't raise, otherwise `False`

**Comment by JJPO96** [comment]
@ThiefMaster What should be the category of the warning message?

**Comment by ThiefMaster** [comment]
maybe RuntimeWarning... I'd check if Flask uses any other warnings (that are not deprecation-related) and possibly use the same type

Also, tests are failing since you don't handle the case where the domain is unset, i.e. `None`

**Comment by JJPO96** [comment]
@ThiefMaster  How can I get the socket of the session when I call the function is_ip() in the save_session() function?

**Comment by ThiefMaster** [comment]
It's a function from python's `socket` module, so you simple `import socket` at the top of the file...

File: flask/sessions.py [hunk 3]
```diff
@@ -332,7 +332,8 @@ def open_session(self, app, request):
 
     def save_session(self, app, session, response):
         domain = self.get_cookie_domain(app)
-        is_IP(domain, "SESSION_COOKIE_DOMAIN")
+        if domain != None:
```
Comment (ThiefMaster) [inline flask/sessions.py]: `if domain is not None:`

File: flask/sessions.py [hunk 4]
```diff
@@ -332,6 +332,8 @@ def open_session(self, app, request):
 
     def save_session(self, app, session, response):
         domain = self.get_cookie_domain(app)
+        if domain != None:
+            is_ip(domain, "SESSION_COOKIE_DOMAIN", self)
```
Comment (ThiefMaster) [inline flask/sessions.py]: Why are you passing `self` here? This doesn't even work. Please test your code (and flask's unittests) before committing/pushing...

**Comment by JJPO96** [comment]
Sorry for that, in my last commit all the unit tests are passing. I also did the changes you asked for.

File: flask/sessions.py [hunk 5]
```diff
@@ -332,6 +333,9 @@ def open_session(self, app, request):
 
     def save_session(self, app, session, response):
         domain = self.get_cookie_domain(app)
+        if domain is not None:
```
Comment (untitaker) [inline flask/sessions.py]: can be conjugated with `and`

File: flask/sessions.py [hunk 6]
```diff
@@ -332,6 +333,9 @@ def open_session(self, app, request):
...
+            if is_ip(domain):
+                warnings.warn("IP introduced in SESSION_COOKIE_DOMAIN", RuntimeWarning)
```
Comment (untitaker) [inline flask/sessions.py]: IMO this should elaborate on why this is a problem

**Comment by davidism** [comment]
@untitaker can you make a decision on this?

**Comment by untitaker** [comment]
I think the code is fine except for my review comments.

**Comment by davidism** [comment]
Setting `SESSION_COOKIE_DOMAIN = '127.0.0.1:5000'` works fine. The issue is when it's detected through `SERVER_NAME` instead, in which case Flask adds a `'.'` before the "domain", which Chrome doesn't like (rightly so, an IP isn't a domain). We can just use the `is_ip` function to decide to remove the `'.'` prefix and things will work as expected.

**Comment by davidism** [comment]
I also don't like the idea of running `is_ip` for essentially every request when ideally `domain` will not be an IP.

**Comment by davidism** [comment]
Continued in #2282.

**Pull request closed without merging**